### PR TITLE
feat(map_loader): add partial map loading interface in pointcloud_map_loader

### PR DESCRIPTION
**Pull request by kminoda**
Signed-off-by: kminoda <koji.minoda@tier4.jp>

## Description
I would like to add a new functionality for some new map loading features in `pointcloud_map_loader`. 
See https://github.com/orgs/autowarefoundation/discussions/2812 for overall idea.
See https://github.com/autowarefoundation/autoware_msgs/tree/main/autoware_map_msgs for the details of each interface.

Note: I will create another PR for `differential_map_loading`, as it will contain some similar implementation as `partial_map_loading`.

## Related links
- https://github.com/autowarefoundation/autoware_msgs/pull/39

## Notes for reviewers
This PR should not have any effect on the current interface (just adding a new interface)

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by kminoda** [comment]
I will add a description in readme and then request for a review

**Comment by kminoda** [comment]
@miursh @Shin-kyoto This is the PR that adds a new interface. (In case you may use this interface in perception modules)

**Comment by kminoda** [comment]
@xmfcx Yes, it is. I would appreciate if you could give us a comment, too!

**Comment by kminoda** [comment]
@RyuYamamoto @yukkysaito This PR is now ready! (also confirmed that the dynamic map loading works on a real vehicle)
Would you review this PR when you have time?

File: map/map_loader/src/pointcloud_map_loader/utils.cpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+     position.z <= position_max.z) ||
+    (position_min.x <= position.x && position.x <= position_max.x &&
+     position_min.y - radius <= position.y && position.y <= position_max.y + radius &&
+     position_min.z <= position.z && position.z <= position_max.z) ||
+    (position_min.x <= position.x && position.x <= position_max.x && position_min.y <= position.y &&
+     position.y <= position_max.y && position_min.z - radius <= position.z &&
+     position.z <= position_max.z + radius)) {
+    return true;
+  }
+  double r2 = std::pow(radius, 2.0);
```
Comment (yukkysaito) [inline map/map_loader/src/pointcloud_map_loader/utils.cpp:35]:
```suggestion
  const double r2 = std::pow(radius, 2.0);
```

File: map/map_loader/src/pointcloud_map_loader/utils.cpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  double minx2 = std::pow(position.x - position_min.x, 2.0);
+  double maxx2 = std::pow(position.x - position_max.x, 2.0);
+  double miny2 = std::pow(position.y - position_min.y, 2.0);
+  double maxy2 = std::pow(position.y - position_max.y, 2.0);
+  double minz2 = std::pow(position.z - position_min.z, 2.0);
+  double maxz2 = std::pow(position.z - position_max.z, 2.0);
```
Comment (yukkysaito) [inline map/map_loader/src/pointcloud_map_loader/utils.cpp:41]: It maybe better to use hypot.
Also, pow is too slow. https://jp.quora.com/C%E3%81%AB%E3%81%8A%E3%81%84%E3%81%A6-%E5%A4%89%E6%95%B0-x-%E3%82%92%E4%BA%8C%E4%B9%97%E3%81%99%E3%82%8B%E5%A0%B4%E5%90%88-pow-x-2-0-%E3%82%88%E3%82%8A-x-x-%E3%81%8C%E6%9C%9B%E3%81%BE%E3%81%97%E3%81%84%E3%81%AE

File: map/map_loader/README.md [hunk 3]
```diff
@@ -6,15 +6,35 @@ This package provides the features of loading various maps.
...
+Given a query from a client node, the node sends a set of pointcloud maps that overlaps with the queried area.
+Please see [the description of `GetPartialPointCloudMap.srv`](https://github.com/autowarefoundation/autoware_msgs/tree/main/autoware_map_msgs#getpartialpointcloudmapsrv) for details.
+
+### Parameters
+
+| Name                | Type   | Description                                                                       | Default value |
+| :------------------ | :----- | :-------------------------------------------------------------------------------- | :------------ |
+| enable_whole_load   | bool   | A flag to enable raw pointcloud map publishing                                    | true          |
+| enable_partial_load | bool   | A flag to enable partial pointcloud map server                                    | true          |
+| leaf_size           | double | Downsampling leaf size (only used when enable_downsampled_whole_load is set true) | 3.0           |
```
Comment (YamatoAndo) [inline map/map_loader/README.md:32]: What is `enable_downsampled_whole_load`?

Comment (kminoda) [reply map/map_loader/README.md]: https://github.com/autowarefoundation/autoware.universe/pull/1938/commits/3bff6ecc2dd2ec2f46cc7a2d3029f5d5e401edc9 Resolved

File: map/map_loader/src/pointcloud_map_loader/utils.cpp (hunk 1 above)
Comment (kminoda) [reply map/map_loader/src/pointcloud_map_loader/utils.cpp]: (removed this line here https://github.com/autowarefoundation/autoware.universe/pull/1938/commits/0cc613ee578104614f53b9876a7409282375ab90)

File: map/map_loader/src/pointcloud_map_loader/utils.cpp (hunk 2 above)
Comment (kminoda) [reply map/map_loader/src/pointcloud_map_loader/utils.cpp]: Thank you for the comment! (and sorry for being late)
I refactored the function using hypot: https://github.com/autowarefoundation/autoware.universe/pull/1938/commits/0cc613ee578104614f53b9876a7409282375ab90

**Comment by kminoda** [comment]
@RyuYamamoto @yukkysaito Friendly ping for your review :eyes:

**Review by yukkysaito — APPROVED** [review]
LGTM

**Review by RyuYamamoto — APPROVED** [review]
Sorry for the late in PR checking.

LGTM

**Comment by kminoda** [comment]
@mitsudome-r @kenji-miyake Do you know why this PR cannot be merged although I got the approve from one of the maintainers (@.RyuYamamoto)?
He's also added in [the CODEOWNERS file](https://github.com/autowarefoundation/autoware.universe/blob/03de605b0998e9a9ee623f9a74a7002f4d2b61a4/.github/CODEOWNERS#L78).

**Comment by kenji-miyake** [comment]
It was due to tier4_map_launch. Added maintainers in https://github.com/autowarefoundation/autoware.universe/pull/2416.

```
launch/tier4_map_launch/** ryohsuke.mitsudome@tier4.jp @autowarefoundation/autoware-global-codeowners
```

**Pull request merged**